src/daft-writers/src/file.rs

```diff
@@ -1,46 +1,80 @@
-use std::sync::Arc;
+use std::{cmp::max, sync::Arc};
 
 use common_error::DaftResult;
 use daft_micropartition::MicroPartition;
 use daft_table::Table;
 
-use crate::{FileWriter, WriterFactory};
+use crate::{FileWriter, TargetInMemorySizeBytesCalculator, WriterFactory};
 
-// TargetFileSizeWriter is a writer that writes in files of a target size.
+// TargetFileSizeWriter is a writer that writes files of a target size.
 // It rotates the writer when the current file reaches the target size.
 struct TargetFileSizeWriter {
-    current_file_rows: usize,
+    current_in_memory_size_estimate: usize,
+    current_bytes_written: usize,
     current_writer: Box<dyn FileWriter<Input = Arc<MicroPartition>, Result = Option<Table>>>,
     writer_factory: Arc<dyn WriterFactory<Input = Arc<MicroPartition>, Result = Option<Table>>>,
-    target_in_memory_file_rows: usize,
+    size_calculator: Arc<TargetInMemorySizeBytesCalculator>,
     results: Vec<Table>,
     partition_values: Option<Table>,
     is_closed: bool,
 }
 
 impl TargetFileSizeWriter {
     fn new(
-        target_in_memory_file_rows: usize,
         writer_factory: Arc<dyn WriterFactory<Input = Arc<MicroPartition>, Result = Option<Table>>>,
         partition_values: Option<Table>,
+        size_calculator: Arc<TargetInMemorySizeBytesCalculator>,
     ) -> DaftResult<Self> {
         let writer: Box<dyn FileWriter<Input = Arc<MicroPartition>, Result = Option<Table>>> =
             writer_factory.create_writer(0, partition_values.as_ref())?;
+        let estimate = size_calculator.calculate_target_in_memory_size_bytes();
         Ok(Self {
-            current_file_rows: 0,
+            current_in_memory_size_estimate: estimate,
+            current_bytes_written: 0,
             current_writer: writer,
             writer_factory,
-            target_in_memory_file_rows,
+            size_calculator,
             results: vec![],
             partition_values,
             is_closed: false,
         })
     }
 
-    fn rotate_writer(&mut self) -> DaftResult<()> {
+    fn remaining_bytes_for_current_file(&self) -> usize {
+        self.current_in_memory_size_estimate
+            .saturating_sub(self.current_bytes_written)
+    }
+
+    fn write_and_update_bytes(
+        &mut self,
+        input: Arc<MicroPartition>,
+        size_bytes: usize,
+    ) -> DaftResult<()> {
+        self.current_bytes_written += size_bytes;
```

Comment:
is this in-memory bytes? If so it could make sense to rename this to `current_in_memory_bytes_written`